### PR TITLE
Fixed #25232 -- Made ModelBackend authentication backend reject inactive users by default.

### PR DESCRIPTION
**Pull request by sasha0**
Ticket - https://code.djangoproject.com/ticket/25232

### CODE REVIEW

File: django/contrib/auth/backends.py [hunk 1]
```diff
@@ -9,13 +9,17 @@ class ModelBackend(object):
...
+    allow_inactive_users = False
+
     def authenticate(self, username=None, password=None, **kwargs):
         UserModel = get_user_model()
         if username is None:
             username = kwargs.get(UserModel.USERNAME_FIELD)
         try:
             user = UserModel._default_manager.get_by_natural_key(username)
-            if user.check_password(password):
+            if user.check_password(password) and \
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: use parenthesis around the expression to avoid backslashes

File: django/contrib/auth/backends.py [hunk 2]
```diff
@@ -9,13 +9,17 @@ class ModelBackend(object):
...
+                    (user.is_active or (not user.is_active and self.allow_inactive_users)):
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: `not user.is_active` not needed.

File: django/contrib/auth/backends.py [hunk 3]
```diff
@@ -9,13 +9,17 @@ class ModelBackend(object):
     Authenticates against settings.AUTH_USER_MODEL.
     """
 
+    # Don't reject user when `is_active` is False.
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: I think the comment says the opposite of the actual behavior.

File: docs/ref/contrib/auth.txt [hunk 4]
```diff
@@ -446,6 +446,11 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
     implement them other than returning an empty set of permissions if
     ``obj is not None``.
 
+    .. attribute:: ModelBackend.allow_inactive_users
+
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: versionadded missing

File: docs/ref/contrib/auth.txt [hunk 5]
```diff
@@ -446,6 +446,11 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
...
+        ``True`` or ``False``. Determines whether or not a
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: chop ``True`` or ``False`` or replace with "Boolean that determines..."
a->the

File: docs/ref/contrib/auth.txt [hunk 6]
```diff
@@ -529,3 +534,14 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
    new user is created, and can be used to perform custom setup actions, such
    as setting the user's groups based on attributes in an LDAP directory.
    Returns the user object.
+
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: one space between sections

File: docs/ref/contrib/auth.txt [hunk 7]
```diff
@@ -529,3 +534,14 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
...
+
+.. class:: AllowInactiveUsersModelBackend
+
+   Same as :class:`~ModelBackend`, but does not reject inactive users by
+   default.
+
+   .. attribute:: AllowInactiveUsersModelBackend.allow_inactive_users
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: I wouldn't list this separately. Above I would say, "doesn't reject inactive users because ~ModelBackend.allow_inactive_users`=True."

File: tests/auth_tests/test_forms.py [hunk 8]
```diff
@@ -154,7 +154,11 @@ def test_validates_password(self):
         )
 
 
-@override_settings(USE_TZ=False, PASSWORD_HASHERS=['django.contrib.auth.hashers.SHA1PasswordHasher'])
+@override_settings(
+    USE_TZ=False,
+    PASSWORD_HASHERS=['django.contrib.auth.hashers.SHA1PasswordHasher'],
+    AUTHENTICATION_BACKENDS=['django.contrib.auth.backends.AllowInactiveUsersModelBackend']
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: include trailing comma in kwargs

File: tests/test_client/tests.py [hunk 9]
```diff
@@ -456,6 +456,9 @@ def test_view_with_inactive_login(self):
         login = self.client.login(username='inactive', password='password')
         self.assertFalse(login)
 
+    @override_settings(AUTHENTICATION_BACKENDS=[
```
Comment (timgraham) [inline tests/test_client/tests.py]: indent style:
```
@override_settings(
    AUTHENTICATION_BACKENDS=[
        'django.contrib.auth.backends.ModelBackend',
        'django.contrib.auth.backends.AllowInactiveUsersModelBackend',
    ],
)
```

File: docs/ref/contrib/auth.txt [hunk 10]
```diff
@@ -529,3 +534,14 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
    new user is created, and can be used to perform custom setup actions, such
    as setting the user's groups based on attributes in an LDAP directory.
    Returns the user object.
+
+
+.. class:: AllowInactiveUsersModelBackend
+
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: versionadded missing

File: django/test/client.py [hunk 11]
```diff
@@ -599,8 +599,7 @@ def login(self, **credentials):
         """
         from django.contrib.auth import authenticate
         user = authenticate(**credentials)
-        if (user and user.is_active and
-                apps.is_installed('django.contrib.sessions')):
+        if user and apps.is_installed('django.contrib.sessions'):
```
Comment (timgraham) [inline django/test/client.py]: Can this change be a separate commit as the fix for ticket #24987?

File: django/contrib/auth/backends.py (hunk 2 above)
Comment (timgraham) [inline django/contrib/auth/backends.py]: We might need to handle the case of custom users that don't have an `is_active` attribute.

**Comment by timgraham** [comment]
Release notes are needed ("backwards-incompatible changes"). I'd be surprised if nothing in `docs/topics/auth` need to be updated. Also update the docs for `.. method:: Client.login(**credentials)` in `docs/topics/testing/tools.txt` (but as noted elsewhere, making that change as a separate pull request to follow this one would make review easier).

File: django/test/client.py (hunk 11 above)
Comment (sasha0) [inline django/test/client.py]: Sure.

File: django/contrib/auth/backends.py (hunk 3 above)
Comment (sasha0) [inline django/contrib/auth/backends.py]: Doh. :confused:

File: django/test/client.py (hunk 11 above)
Comment (sasha0) [inline django/test/client.py]: Done. Most of all you could suggest better explaining commit message.

File: django/contrib/auth/backends.py [hunk 12]
```diff
@@ -9,13 +9,17 @@ class ModelBackend(object):
...
+
     def authenticate(self, username=None, password=None, **kwargs):
         UserModel = get_user_model()
         if username is None:
             username = kwargs.get(UserModel.USERNAME_FIELD)
         try:
             user = UserModel._default_manager.get_by_natural_key(username)
-            if user.check_password(password):
+            if (user.check_password(password) and
+                    (user.is_active or (not user.is_active and self.allow_inactive_users))):
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: same thing here about assuming `is_active` exists and `not user.is_active` -- probably need some tests for that case.

File: django/contrib/auth/backends.py [hunk 13]
```diff
@@ -90,9 +94,12 @@ def has_module_perms(self, user_obj, app_label):
     def get_user(self, user_id):
         UserModel = get_user_model()
         try:
-            return UserModel._default_manager.get(pk=user_id)
+            user = UserModel._default_manager.get(pk=user_id)
         except UserModel.DoesNotExist:
             return None
+        if user and (getattr(user, 'is_active', None) or self.allow_inactive_users):
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: I tend to think of the user doesn't have an `is_active` attribute, it should be accepted regardless of `allow_inactive_users`. Should make the behavior clear in the docs.

File: docs/ref/contrib/auth.txt [hunk 14]
```diff
@@ -446,6 +446,15 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
     implement them other than returning an empty set of permissions if
     ``obj is not None``.
 
+    .. attribute:: ModelBackend.allow_inactive_users
+
+        Boolean that determines whether or not the
+        backend will reject inactive users. Defaults to ``False``.
+
+        .. versionchanged:: 1.10
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: Just put `.. versionchanged:: 1.10` between the `.. attribute::`line and the paragraph.

File: docs/ref/contrib/auth.txt [hunk 15]
```diff
@@ -529,3 +538,8 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
    new user is created, and can be used to perform custom setup actions, such
    as setting the user's groups based on attributes in an LDAP directory.
    Returns the user object.
+
+.. class:: AllowInactiveUsersModelBackend
+
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: versionadded

File: docs/ref/contrib/auth.txt [hunk 16]
```diff
@@ -529,3 +536,10 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
...
+   .. versionchanged:: 1.10
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: versionadded, not changed

File: docs/ref/contrib/auth.txt [hunk 17]
```diff
@@ -446,6 +446,13 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
     implement them other than returning an empty set of permissions if
     ``obj is not None``.
 
+    .. attribute:: ModelBackend.allow_inactive_users
+
+        .. versionchanged:: 1.10
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: versionadded, not changed

File: docs/ref/contrib/auth.txt [hunk 18]
```diff
@@ -446,6 +446,13 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
...
+        
+            Boolean that determines whether or not the
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: deindent

File: docs/ref/contrib/auth.txt [hunk 19]
```diff
@@ -529,3 +536,10 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
    new user is created, and can be used to perform custom setup actions, such
    as setting the user's groups based on attributes in an LDAP directory.
    Returns the user object.
+
+.. class:: AllowInactiveUsersModelBackend
+
+   .. versionchanged:: 1.10
+
+       Same as :class:`~ModelBackend`, but doesn't reject inactive users because
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: deindent

**Comment by sasha0** [comment]
> Release notes are needed ("backwards-incompatible changes"). I'd be surprised if nothing in docs/topics/auth need to be updated. Also update the docs for .. method:: Client.login(**credentials) in docs/topics/testing/tools.txt (but as noted elsewhere, making that change as a separate pull request to follow this one would make review easier).

Thanks for the suggestion, working on it.

So you think it's better to create a separate PR with documentation changes? If no, please clarify what changes specifically you'd recommend to move to the separate PR. Thanks!

**Comment by timgraham** [comment]
I meant the entire patch for #24987 as a separate pull request.

**Comment by sasha0** [comment]
Good, will do.

File: django/contrib/auth/backends.py (hunk 12 above)
Comment (sasha0) [inline django/contrib/auth/backends.py]: Test to check whether user without `is_active` field will be allowed to login, right?

Comment (timgraham) [inline django/contrib/auth/backends.py]: Correct

Comment (sasha0) [inline django/contrib/auth/backends.py]: Ok, added test.
Also, added tests for `django.contrib.auth.backends.AllowInactiveUsersModelBackendTest`.
Was also thinking about inheriting test suite from `BaseModelTests` but decided that it's overhead.

Please let me know what do you think.

File: django/contrib/auth/backends.py (hunk 13 above)
Comment (sasha0) [inline django/contrib/auth/backends.py]: Done. Made this check same in `authenticate` and `get_user`.

File: tests/auth_tests/test_auth_backends.py [hunk 20]
```diff
@@ -667,3 +683,30 @@ def test_backend_path_login_with_explicit_backends(self):
...
+class AllowInactiveUsersModelBackendTest(TestCase):
+
+    def setUp(self):
+        self.user = User.objects.create_user('test', 'test@example.com', 'test')
+        self.user.is_active = False
+        self.user.save()
+
+    def test_authenticate(self):
+        self.assertEqual(authenticate(username='test', password='test'), self.user)
+
```
Comment (sasha0) [inline tests/auth_tests/test_auth_backends.py]: @timgraham this test won't pass until #6124 is merged, because inactive user is rejected by test client. What would you suggest in this case? Thanks.

Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: We can move the test that involves `login()` to the other pull request.

File: django/contrib/auth/backends.py [hunk 21]
```diff
@@ -90,9 +96,15 @@ def has_module_perms(self, user_obj, app_label):
     def get_user(self, user_id):
         UserModel = get_user_model()
         try:
-            return UserModel._default_manager.get(pk=user_id)
+            user = UserModel._default_manager.get(pk=user_id)
         except UserModel.DoesNotExist:
             return None
+        if user:
+            is_active = getattr(user, 'is_active', None)
+            if (is_active or (is_active is False and self.allow_inactive_users) or
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: Isn't `is_active is False` redundant? We already check if it's truthy in the first condition so if we get here it must be falsey.

File: docs/ref/contrib/auth.txt [hunk 22]
```diff
@@ -446,6 +446,13 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
     implement them other than returning an empty set of permissions if
     ``obj is not None``.
 
+    .. attribute:: ModelBackend.allow_inactive_users
+
+        .. versionadded:: 1.10
+        
+        Boolean that determines whether or not the
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: wrap this line closer to 79 characters

File: docs/releases/1.10.txt [hunk 23]
```diff
@@ -552,6 +552,9 @@ Miscellaneous
 
 * Tests that violate deferrable database constraints will now error when run on
   a database that supports deferrable constraints.
+* ``django.contrib.auth.backends.ModelBackend`` now rejects inactive users by
+  default. If you need previous behaviour, use
```
Comment (timgraham) [inline docs/releases/1.10.txt]: the previous behavior
+ add blank line above the item

File: docs/topics/auth/customizing.txt [hunk 24]
```diff
@@ -234,10 +234,12 @@ for example, to control anonymous access.
...
-An inactive user is a one that is authenticated but has its attribute
-``is_active`` set to ``False``. However this does not mean they are not
-authorized to do anything. For example they are allowed to activate their
-account.
+An inactive user is a one that has its attribute ``is_active`` set to
+``False``. By default, they can't authorize until their accounts will be
+activated. To change this, you need to override
+:meth:`~django.contrib.auth.backends.ModelBackend.authenticate` and
+:meth:`~django.contrib.auth.backends.ModelBackend.get_user`.
+Also, if user has no ``is_active`` field - he will be allowed to login.
```
Comment (timgraham) [inline docs/topics/auth/customizing.txt]: he->they (we cannot use gendered language in the docs)
Looks like "Also, ..." could be wrapped to the previous line.

File: tests/auth_tests/models/custom_user.py [hunk 25]
```diff
@@ -95,6 +97,34 @@ def __exit__(self, exc_type, exc_value, traceback):
         PermissionsMixin._meta.local_many_to_many = self._old_pm_local_m2m
 
 
+class CustomUserWithoutIsActiveField(AbstractBaseUser):
+    username = models.CharField(
+        _('username'),
+        max_length=150,
+        unique=True,
+        help_text=_('Required. 150 characters or fewer. Letters, digits and @/./+/-/_ only.'),
+        validators=[
```
Comment (timgraham) [inline tests/auth_tests/models/custom_user.py]: try to make this as minimal as possible -- if we don't need all the validators, error_messages, etc. leave them out

File: tests/auth_tests/models/custom_user.py [hunk 26]
```diff
@@ -95,6 +97,34 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+        },
+    )
+    email = models.EmailField(verbose_name='email address', max_length=255, unique=True)
+
+    objects = UserManager()
+
+    USERNAME_FIELD = 'username'
+    REQUIRED_FIELDS = ['email']
+
+    def __unicode__(self):
```
Comment (timgraham) [inline tests/auth_tests/models/custom_user.py]: I guess the method probably isn't needed but if you want to keep it, please fix it like c62807968d7930bfd34afc2036c67921b943592f.

File: tests/auth_tests/models/custom_user.py [hunk 27]
```diff
@@ -95,6 +97,34 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+                r'^[\w.@+-]+$',
+                _('Enter a valid username. This value may contain only '
+                  'letters, numbers ' 'and @/./+/-/_ characters.')
+            ),
+        ],
+        error_messages={
+            'unique': _("A user with that username already exists."),
+        },
+    )
+    email = models.EmailField(verbose_name='email address', max_length=255, unique=True)
```
Comment (timgraham) [inline tests/auth_tests/models/custom_user.py]: do we need the email field for the purpose of the test?

File: tests/auth_tests/test_auth_backends.py [hunk 28]
```diff
@@ -194,6 +197,19 @@ def test_authentication_timing(self):
...
         self.assertEqual(CountingMD5PasswordHasher.calls, 1)
 
+    def test_authenticate_inactive(self):
+        # Inactive user is rejected by default
+        user = self.UserModel._default_manager.get(pk=self.user.pk)
+        user.is_active = False
+        user.save()
+        self.assertIsNone(authenticate(username='test', password='test'))
+        # User without `is_active` field is allowed
+        with override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithoutIsActiveField'):
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: I'd make this a separate test and then you can decorator the test method instead of using `with ....`.

File: tests/auth_tests/test_auth_backends.py [hunk 29]
```diff
@@ -667,3 +683,30 @@ def test_backend_path_login_with_explicit_backends(self):
...
         self.client._login(user, self.other_backend)
         self.assertBackendInSession(self.other_backend)
+
+
+@override_settings(
+    AUTHENTICATION_BACKENDS=['django.contrib.auth.backends.AllowInactiveUsersModelBackend'],
+)
+class AllowInactiveUsersModelBackendTest(TestCase):
+
+    def setUp(self):
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: prefer `setUpTestData` since that executes once per test class instead of once for every method

File: tests/auth_tests/test_forms.py [hunk 30]
```diff
@@ -139,7 +139,9 @@ class Meta(UserCreationForm.Meta):
         form = CustomUserCreationForm(data)
         self.assertTrue(form.is_valid())
 
-
+@override_settings(
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: It would be helpful to add an explanatory comment like "To verify that the login form rejects inactive users, use an authentication backend that allows them." Something similar for the other places you modified might be useful as well.

File: django/contrib/auth/backends.py (hunk 21 above)
Comment (sasha0) [inline django/contrib/auth/backends.py]: I think it's not - it might be `False` or `None` and we return user either account is inactive and backend allows it, or account does not have this field. What do you think?

Comment (timgraham) [inline django/contrib/auth/backends.py]: What values would this condition have a different outcome than the current one?
`is_active or self.allow_inactive_users or is_active is None`

File: tests/auth_tests/test_auth_backends.py (hunk 28 above)
Comment (sasha0) [inline tests/auth_tests/test_auth_backends.py]: Ok. Wouldn't it also better to move test to `ModelBackendTest` class and not keep in the base test class, so it won't be executed within every child test class.

File: django/contrib/auth/backends.py (hunk 21 above)
Comment (sasha0) [inline django/contrib/auth/backends.py]: Ah, right, this one should give same result. Thanks for the suggestion.

File: docs/topics/auth/customizing.txt (hunk 24 above)
Comment (sasha0) [inline docs/topics/auth/customizing.txt]: Ok. Also `django.contrib.auth.backends.ModelBackend.get_user` is not documented yet thus it gives a warning during `make spelling`. What would you suggest: avoid to use this reference or to document `get_user` method? Thanks.

Comment (timgraham) [inline docs/topics/auth/customizing.txt]: Shouldn't we recommend use of `AllowInactiveUsersModelBackend` instead of overriding those methods?

File: tests/auth_tests/test_auth_backends.py (hunk 28 above)
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: sure

File: docs/topics/auth/customizing.txt (hunk 24 above)
Comment (sasha0) [inline docs/topics/auth/customizing.txt]: Indeed. Thanks!

File: tests/auth_tests/models/custom_user.py (hunk 27 above)
Comment (sasha0) [inline tests/auth_tests/models/custom_user.py]: No, if we create user without `User.objects.create_user` method, because it requires it.

File: tests/auth_tests/test_auth_backends.py (hunk 20 above)
Comment (sasha0) [inline tests/auth_tests/test_auth_backends.py]: Ok done, move to #6201.

**Comment by timgraham** [comment]
Do you think these edits are an improvement? http://dpaste.com/13EEX4Q

File: docs/ref/contrib/auth.txt [hunk 31]
```diff
@@ -446,6 +446,13 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
     implement them other than returning an empty set of permissions if
     ``obj is not None``.
 
+    .. attribute:: ModelBackend.allow_inactive_users
+
+        .. versionadded:: 1.10
+
```
Comment (timgraham) [inline docs/ref/contrib/auth.txt]: remove whitespace on blank line (trying configuring your editor to remove trailing whitespace on save)

File: docs/releases/1.10.txt [hunk 32]
```diff
@@ -617,6 +617,10 @@ Miscellaneous
 * Tests that violate deferrable database constraints will now error when run on
   a database that supports deferrable constraints.
 
+* ``django.contrib.auth.backends.ModelBackend`` now rejects inactive users by
```
Comment (timgraham) [inline docs/releases/1.10.txt]: use links to use classes?

File: docs/releases/1.10.txt [hunk 33]
```diff
@@ -617,6 +617,10 @@ Miscellaneous
...
+  default. If you need previous behavior, use
```
Comment (timgraham) [inline docs/releases/1.10.txt]: "the previous"

File: tests/auth_tests/models/custom_user.py [hunk 34]
```diff
@@ -97,6 +98,19 @@ def __exit__(self, exc_type, exc_value, traceback):
         PermissionsMixin._meta.local_many_to_many = self._old_pm_local_m2m
 
 
+class CustomUserWithoutIsActiveField(AbstractBaseUser):
+    username = models.CharField(
```
Comment (timgraham) [inline tests/auth_tests/models/custom_user.py]: single line here should be fine

File: tests/auth_tests/test_auth_backends.py [hunk 35]
```diff
@@ -213,6 +216,31 @@ def create_users(self):
             password='test',
         )
 
+    def test_authenticate_inactive(self):
+        """
+        To verify that inactive user being rejected.
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: State the expected behavior, "An active user can't authenticate."

File: tests/auth_tests/test_auth_backends.py [hunk 36]
```diff
@@ -213,6 +216,31 @@ def create_users(self):
...
+        self.user.save()
+        self.assertIsNone(
+            authenticate(
+                username='test',
+                email='test@example.com',
+                password='test'
+            )
+        )
+
+    # To make sure user without `is_active` field is allowed to
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: as above, state expected behavior in a docstring

File: tests/auth_tests/test_auth_backends.py [hunk 37]
```diff
@@ -213,6 +216,31 @@ def create_users(self):
             password='test',
         )
 
+    def test_authenticate_inactive(self):
+        """
+        To verify that inactive user being rejected.
+        """
+        self.user.is_active = False
+        self.user.save()
+        self.assertIsNone(
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: single line please

File: tests/auth_tests/test_auth_backends.py [hunk 38]
```diff
@@ -213,6 +216,31 @@ def create_users(self):
...
+            authenticate(
+                username='test',
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: to ensure the a typo in the credentials isn't the reason for the failure, you could create a dict with them and verify authentication is successful before you set `self.user.is_active = False`. e.g. `self.assertEqual(authenticate(**credentials), self.user)`

File: tests/auth_tests/test_auth_backends.py [hunk 39]
```diff
@@ -667,3 +695,43 @@ def test_backend_path_login_with_explicit_backends(self):
...
+                email='test@example.com',
+                password='test'
+            ),
+            self.user
+        )
+
+    def test_get_user(self):
+        self.client.force_login(self.user)
+
+        # Prepare a request object
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: unnecessary comment

File: tests/auth_tests/test_auth_backends.py [hunk 40]
```diff
@@ -667,3 +695,43 @@ def test_backend_path_login_with_explicit_backends(self):
...
+    To verify that the login form rejects inactive users, use an
+    authentication backend that allows them.
+    """
+    @classmethod
+    def setUpTestData(cls):
+        cls.user = User.objects.create_user(
+            username='test',
+            email='test@example.com',
+            password='test',
+            is_active=False
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: add trailing comma

File: tests/auth_tests/test_auth_backends.py [hunk 41]
```diff
@@ -667,3 +695,43 @@ def test_backend_path_login_with_explicit_backends(self):
...
+)
+class AllowInactiveUsersModelBackendTest(TestCase):
+    """
+    To verify that the login form rejects inactive users, use an
+    authentication backend that allows them.
+    """
+    @classmethod
+    def setUpTestData(cls):
+        cls.user = User.objects.create_user(
+            username='test',
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: store credentials in dict for easy reuse as suggested above?

File: tests/auth_tests/test_forms.py [hunk 42]
```diff
@@ -140,6 +140,9 @@ class Meta(UserCreationForm.Meta):
         self.assertTrue(form.is_valid())
 
 
+@override_settings(
+    AUTHENTICATION_BACKENDS=['django.contrib.auth.backends.AllowInactiveUsersModelBackend'],
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: Add comment about why this is necessary.

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/6309.
